File: app/src/main/java/com/bookcrossing/mobile/util/SystemServicesWrapper.java
package com.bookcrossing.mobile.util;

import android.content.SharedPreferences;
import com.bookcrossing.mobile.modules.App;
import dagger.Lazy;
import io.nlopez.smartlocation.SmartLocation;
import javax.inject.Inject;

/**
 * Created by fobo66 on 25.4.17.
 */

public class SystemServicesWrapper {
  @Inject Lazy<SharedPreferences> preferencesLazy;

  @Inject Lazy<App> appLazy;

  @Inject Lazy<SmartLocation> locationLazy;

  public SharedPreferences getPreferences() {
    return preferencesLazy.get();
  }

  public App getApp() {
    return appLazy.get();
  }

  public SmartLocation getLocation() {
    return locationLazy.get();
  }
}

File: app/src/main/java/com/bookcrossing/mobile/util/SystemServicesWrapper.kt
package com.bookcrossing.mobile.util

import android.content.SharedPreferences

import com.bookcrossing.mobile.location.LocationRepository
import com.bookcrossing.mobile.modules.App
import dagger.Lazy
import io.nlopez.smartlocation.SmartLocation
import javax.inject.Inject

/**
 * Created by fobo66 on 25.4.17.
 */

class SystemServicesWrapper @Inject constructor(
        private val preferencesLazy: Lazy<SharedPreferences>,
        private val appLazy: Lazy<App>,
        private val locationLazy: Lazy<SmartLocation>,
        private val locationRepositoryLazy: Lazy<LocationRepository>
) {


    val preferences: SharedPreferences
        get() = preferencesLazy.get()

    val app: App
        get() = appLazy.get()

    val location: SmartLocation
        get() = locationLazy.get()

    val locationRepository: LocationRepository
        get() = locationRepositoryLazy.get()
}
